pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -4,67 +4,243 @@ package querymiddleware
 
 import (
 	"bytes"
+	"context"
 	"fmt"
 	"io"
 	"net/http"
 	"net/url"
 	"strconv"
 	"strings"
 
+	"github.com/opentracing/opentracing-go"
 	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/promql/parser"
 	"github.com/prometheus/prometheus/storage/remote"
 
+	apierror "github.com/grafana/mimir/pkg/api/error"
 	"github.com/grafana/mimir/pkg/querier"
 	"github.com/grafana/mimir/pkg/util"
 )
 
+// To keep logs and error messages in sync, we define the following keys:
+const (
+	endLogKey      = "end"
+	hintsLogKey    = "hints"
+	matchersLogKey = "matchers"
+	startLogKey    = "start"
+)
+
+type remoteReadRoundTripper struct {
+	next http.RoundTripper
+
+	middleware MetricsQueryMiddleware
+}
+
+func newRemoteReadRoundTripper(next http.RoundTripper, middlewares ...MetricsQueryMiddleware) http.RoundTripper {
+	return &remoteReadRoundTripper{
+		next:       next,
+		middleware: MergeMetricsQueryMiddlewares(middlewares...),
+	}
+}
+
+func (r *remoteReadRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	remoteReadRequest, err := getRemoteReadRequestWithoutConsumingBody(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if remoteReadRequest == nil {
+		return r.next.RoundTrip(req)
+	}
+
+	queries := remoteReadRequest.GetQueries()
+	for i, query := range queries {
+		metricsRequest, err := toMetricsRequest(req.URL.Path, query)
+		if err != nil {
+			return nil, err
+		}
+		handler := r.middleware.Wrap(HandlerFunc(func(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+			// We do not need to do anything here as this middleware is used for
+			// validation only and previous middlewares would have already returned errors.
+			return nil, nil
+		}))
+		_, error := handler.Do(req.Context(), metricsRequest)
```

Comment:
[nit] `error` is the global data type. I suggest to get stick to the `err` convention.
```suggestion
		_, err := handler.Do(req.Context(), metricsRequest)
```

Reply:
facepalm